{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6022bf07-3e2b-4f4e-b3c7-194f062377e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "inFile = \"/home/ross/rossgit/holoviews-examples/huge.paf\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "005e3f52-f2f1-4a4d-8847-05f6d1fb19b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# run this to install all required dependencies before running the main code if not already in the venv\n",
    "! pip install datashader dask[dataframe] holoviews[recommended] pandas matplotlib bokeh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c2884a0b-ed41-4c60-ad5f-ee61b87c0c77",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# RUN THIS to create three interactive HiC displays in a Jypyter notebook\n",
    "# after loading dependencies with\n",
    "# ! pip install datashader dask[dataframe] holoviews[recommended] pandas matplotlib bokeh\n",
    "# The data are Arima HiC reads from the Arctic Ground Squirrel mUroPar1 VGP genomeArk repository\n",
    "# processed with Dephine's Pretext workflow using Bellerophon to remove chimeric reads\n",
    "# The paired bam output is converted to PAF with an awk script (!) and that's what is read\n",
    "# in this code.\n",
    "# The pairs are parsed to extract the haplotype designator from the contig name\n",
    "# typically a suffix like H1 extracted in getHap - rewrite that to suit your names.\n",
    "# Sorting by contig name is based on VGP conventions - SUPER_ first, then scaffolds\n",
    "# One problem to watch out for is that any differences in ordering of the X and Y contigs can make all sorts of\n",
    "# artifacts appear such as the kaleidoscopic patterns seen in Pretextviewer.\n",
    "# Ross Lazarus October 2024\n",
    "\n",
    "from bisect import bisect_left\n",
    "from collections import OrderedDict\n",
    "from functools import cmp_to_key\n",
    "import math\n",
    "import numpy as np\n",
    "\n",
    "import holoviews as hv\n",
    "import panel as pn\n",
    "import pandas as pd\n",
    "\n",
    "from holoviews.operation.datashader import (\n",
    "    rasterize,\n",
    "    dynspread,\n",
    ")\n",
    "from holoviews.operation.resample import ResampleOperation2D\n",
    "from holoviews.operation import decimate\n",
    "\n",
    "useDecimate = False # will rasterize instead\n",
    "# inFile = \"galaxy_inputs/paf/bothmap.paf.tab.tabular\"\n",
    "inFile = \"/home/ross/rossgit/holoviews-examples/huge.paf\"\n",
    "ptwidth = 800\n",
    "pcwidth = 800\n",
    "# width settings for plots and location bars\n",
    "hv.extension(\"bokeh\")\n",
    "pn.extension()\n",
    "\n",
    "def rotatecoords(x, y, radians=0.7853981633974483, origin=(0, 0)):\n",
    "    # https://gist.github.com/LyleScott/d17e9d314fbe6fc29767d8c5c029c362\n",
    "    offset_x, offset_y = origin\n",
    "    adjusted_x = x - offset_x\n",
    "    adjusted_y = y - offset_y\n",
    "    cos_rad = math.cos(radians)\n",
    "    sin_rad = math.sin(radians)\n",
    "    qx = offset_x + cos_rad * adjusted_x + sin_rad * adjusted_y\n",
    "    qy = offset_y + -sin_rad * adjusted_x + cos_rad * adjusted_y\n",
    "    return qx, qy\n",
    "\n",
    "\n",
    "def getHap(contig):\n",
    "    \"\"\"\n",
    "    function to return suffix H1 from chrH1 - adjust to suit.\n",
    "    \"\"\"\n",
    "    return contig[-2:]\n",
    "\n",
    "\n",
    "def xportHtml():\n",
    "    \"save a holoview object to an interactive but not adaptive scaling HTML page\"\n",
    "    p = (\n",
    "        rasterize(pafcis1)\n",
    "        .relabel(\"%s Cis HiC interactions\" % haps[0])\n",
    "        .opts(\n",
    "            cmap=\"inferno\",\n",
    "            cnorm=\"log\",\n",
    "            colorbar=True,\n",
    "            width=1000,\n",
    "            height=600,\n",
    "            xticks=tic1,\n",
    "            yticks=tic1,\n",
    "            xrotation=45,\n",
    "            fontsize={\"xticks\": 5, \"yticks\": 5},\n",
    "        )\n",
    "    )\n",
    "    hv_pane = pn.pane.HoloViews(p, height=2000, width=2000)\n",
    "    hv.save(filename=\"H1cis.html\", obj=hv_pane)\n",
    "\n",
    "\n",
    "def sorthapqname(s1, s2):\n",
    "    \"\"\"\n",
    "    fugly hack to sort super contigs before anything else\n",
    "    then by contig number or if they are the same offset\n",
    "    ('SUPER_2H1', 226668729) , ('SUPER_1H1', 284260672), ('SUPER13_unloc_5H1',..), (Scaffold_1116H2, ...)\n",
    "    \"\"\"\n",
    "    if s1[0] == s2[0]:  # simplest case - same contig, sort on offset\n",
    "        return s1[1] - s2[1]  # neg if left sorts before\n",
    "    s11, s12 = s1[0].split(\"_\", 1)\n",
    "    s1n = s12.split(\"_\")[-1][:-2]\n",
    "    s1_super = (s11.upper() == \"SUPER\") or (s11.upper().startswith(\"CHR\"))\n",
    "    s21, s22 = s2[0].split(\"_\", 1)\n",
    "    s2n = s22.split(\"_\")[-1][:-2]\n",
    "    s2_super = (s21.upper() == \"SUPER\") or (s21.upper().startswith(\"CHR\"))\n",
    "    if s1n.isdigit():\n",
    "        s1nn = int(s1n)\n",
    "    else:\n",
    "        s1nn = ord(s1n[0]) * 1000\n",
    "    if s2n.isdigit():\n",
    "        s2nn = int(s2n)\n",
    "    else:\n",
    "        s2nn = ord(s2n[0]) * 1000\n",
    "    if s1_super == s2_super:\n",
    "        nunder1 = len(s1[0].split(\"_\"))\n",
    "        nunder2 = len(s2[0].split(\"_\"))  # _unloc or whatever\n",
    "        if nunder1 == nunder2:\n",
    "            return s1nn - s2nn\n",
    "        else:\n",
    "            return nunder1 - nunder2\n",
    "    elif s1_super:\n",
    "        return -1\n",
    "    elif s2_super:\n",
    "        return 1\n",
    "    else:\n",
    "        return s1nn - s2nn\n",
    "\n",
    "\n",
    "def showH1(x, y):\n",
    "    if np.isnan(x) or np.isnan(y):\n",
    "        s = 'Mouse click on image for location'\n",
    "    else:\n",
    "        i = bisect_left(h1starts, x)\n",
    "        chrx = h1names[i-1]\n",
    "        offsx = x - h1starts[i-1]\n",
    "        i = bisect_left(h1starts, y)\n",
    "        chry = h1names[i-1]\n",
    "        offsy = y - h1starts[i-1]        \n",
    "        s ='X axis %s:%d Y axis %s:%d' % (chrx,offsx, chry, offsy)\n",
    "    str_pane = pn.pane.Str(s, styles={'font-size': '10pt', 'color': 'darkblue', 'text-align': 'center'}, width=pcwidth)\n",
    "    return str_pane\n",
    "\n",
    "\n",
    "def showH2(x, y):\n",
    "    if np.isnan(x) or np.isnan(y):\n",
    "        s = 'Mouse click on image for location'\n",
    "    else:\n",
    "        i = bisect_left(h2starts, x)\n",
    "        chrx = h2names[i-1]\n",
    "        offsx = x - h2starts[i-1]\n",
    "        i = bisect_left(h2starts, y)\n",
    "        chry = h2names[i-1]\n",
    "        offsy = y - h2starts[i-1]        \n",
    "        s ='X axis %s:%d Y axis %s:%d' % (chrx,offsx, chry, offsy)\n",
    "    str_pane = pn.pane.Str(s, styles={'font-size': '11pt', 'color': 'darkblue', 'text-align': 'center'}, width=pcwidth)\n",
    "    return str_pane\n",
    "\n",
    "def showTrans(x, y):\n",
    "    if np.isnan(x) or np.isnan(y):\n",
    "        s = 'Mouse click on image for location'\n",
    "    else:\n",
    "        i = bisect_left(h1starts, x)\n",
    "        chrx = h1names[i-1]\n",
    "        offsx = x - h1starts[i-1]\n",
    "        i = bisect_left(h2starts, y)\n",
    "        chry = h2names[i-1]\n",
    "        offsy = y - h2starts[i-1]        \n",
    "        s ='X axis %s:%d Y axis %s:%d' % (chrx,offsx, chry, offsy)\n",
    "    str_pane = pn.pane.Str(s, styles={'font-size': '11pt', 'color': 'darkblue', 'text-align': 'center'}, width=pcwidth)\n",
    "    return str_pane\n",
    "\n",
    "\n",
    "\n",
    "# Default values suitable for this notebook\n",
    "decimate.max_samples = 10000\n",
    "dynspread.max_px = 10\n",
    "dynspread.threshold = 0.6\n",
    "ResampleOperation2D.width = pcwidth\n",
    "ResampleOperation2D.height = pcwidth\n",
    "# need to convert the categorical contigs into a sequence for holoviews to munch\n",
    "# use the contig length from the paf to figure out the cumulative start for each contig\n",
    "# contigs are length ordered - that does not always work well when the haplotypes differ widely\n",
    "\n",
    "hlstarts = OrderedDict()\n",
    "hqstarts = OrderedDict()\n",
    "hlens = {}\n",
    "haps = []\n",
    "hlsorts = {}\n",
    "hqsorts = {}\n",
    "with open(inFile, \"r\") as f:\n",
    "    for i, rows in enumerate(f):\n",
    "        row = rows.split()\n",
    "        c1 = row[0]\n",
    "        c2 = row[5]\n",
    "        hap = getHap(c1)\n",
    "        if hap not in haps:\n",
    "            haps.append(hap)\n",
    "            hlens[hap] = {}\n",
    "            hlsorts[hap] = []\n",
    "            hqsorts[hap] = []\n",
    "        if not hlens[hap].get(c1, None):\n",
    "            hlens[hap][c1] = int(row[1])\n",
    "            hlsorts[hap].append((int(row[1]), c1))\n",
    "            hqsorts[hap].append((c1, int(row[1])))\n",
    "for hap in haps:\n",
    "    cum = 1\n",
    "    hlsorts[hap].sort(reverse=True)\n",
    "    hqsorts[hap].sort(key=cmp_to_key(sorthapqname))\n",
    "    hlstarts[hap] = OrderedDict()\n",
    "    hqstarts[hap] = OrderedDict()\n",
    "    for clen, contig in hlsorts[hap]:\n",
    "        hlstarts[hap][contig] = cum\n",
    "        cum += clen\n",
    "    cum = 1\n",
    "    for contig, clen in hqsorts[hap]:\n",
    "        hqstarts[hap][contig] = cum\n",
    "        cum += clen\n",
    "h1starts = [hqstarts[haps[0]][x] for x in hqstarts[haps[0]].keys()]\n",
    "h1names = list(hqstarts[haps[0]].keys())\n",
    "h2starts = [hqstarts[haps[1]][x] for x in hqstarts[haps[1]].keys()]\n",
    "h2names = list(hqstarts[haps[1]].keys())\n",
    "# have the axes set up so prepare the three plot x/y vectors\n",
    "# for a second pass to calculate all the coordinates.\n",
    "# adding tooltips just does not scale so abando - see the tooltip old version\n",
    "cis1 = {\"x\":[], \"y\":[]}\n",
    "cis2 = {\"x\":[], \"y\":[]}\n",
    "trans1 = {\"x\":[], \"y\":[]}\n",
    "with open(inFile, \"r\") as f:\n",
    "    for rows in f:\n",
    "        row = rows.split()\n",
    "        c1 = row[0]\n",
    "        c2 = row[5]\n",
    "        H1 = getHap(c1)\n",
    "        H2 = getHap(c2)\n",
    "        if H1 != H2:  # trans\n",
    "            if H1 == haps[0]:  # x is h1 for trans - otherwise ignore\n",
    "                trans1['x'].append(hqstarts[H1][c1] + int(row[2]))\n",
    "                trans1['y'].append(hqstarts[H2][c2] + int(row[7]))\n",
    "            else:\n",
    "                trans1['y'].append(hqstarts[H1][c1] + int(row[2]))\n",
    "                trans1['x'].append(hqstarts[H2][c2] + int(row[7]))\n",
    "        else:  # cis\n",
    "            if H1 == haps[0]:                \n",
    "                cis1['x'].append(hqstarts[H1][c1] + int(row[2]))\n",
    "                cis1['y'].append(hqstarts[H2][c2] + int(row[7]))\n",
    "            else:               \n",
    "                cis2['x'].append(hqstarts[H1][c1] + int(row[2]))\n",
    "                cis2['y'].append(hqstarts[H2][c2] + int(row[7]))\n",
    "hap = haps[0]\n",
    "qtic1 = [(hqstarts[hap][x], x) for x in hqstarts[hap].keys()]\n",
    "hap = haps[1]\n",
    "qtic2 = [(hqstarts[hap][x], x) for x in hqstarts[hap].keys()]\n",
    "# to rotate so the diagonal becomes the x axis\n",
    "# xcis1r, ycis1r = rotatecoords(xcis1, ycis1, radians=0.7853981633974483, origin=(max(xcis1),max(ycis1)))\n",
    "# pafxycis1 = pd.DataFrame(np.vstack([xcis1r,ycis1r]).T, columns = ['x', 'y'])\n",
    "# bisect.bisect_left(a, x, lo=0, hi=len(a), *, key=None)\n",
    "pafxycis1 = pd.DataFrame.from_dict(cis1)\n",
    "pafxycis2 = pd.DataFrame.from_dict(cis2)\n",
    "pafxytrans = pd.DataFrame.from_dict(trans1)\n",
    "pafcis1 = hv.Points(pafxycis1)\n",
    "pafcis2 = hv.Points(pafxycis2)\n",
    "paftrans = hv.Points(pafxytrans)\n",
    "streamcis1 = hv.streams.Tap(source=pafcis1, x=np.nan, y=np.nan)\n",
    "streamcis2 = hv.streams.Tap(source=pafcis2, x=np.nan, y=np.nan)\n",
    "streamtrans = hv.streams.Tap(source=paftrans, x=np.nan, y=np.nan)\n",
    "showloc1 = pn.bind(showH1, x=streamcis1.param.x, y=streamcis1.param.y)\n",
    "showloc2 = pn.bind(showH2, x=streamcis2.param.x, y=streamcis2.param.y)\n",
    "showloctrans = pn.bind(showTrans, x=streamtrans.param.x, y=streamtrans.param.y)\n",
    "p1 = pn.Column(showloc1, pn.pane.HoloViews(dynspread(rasterize(pafcis1).relabel(\"%s Cis HiC interactions\" % haps[0]).opts(\n",
    "            cmap=\"inferno\",\n",
    "            cnorm=\"log\",\n",
    "            colorbar=True,\n",
    "            width=pcwidth,\n",
    "            height=pcwidth,\n",
    "            xticks=qtic1,\n",
    "            yticks=qtic1,\n",
    "            xrotation=45,\n",
    "            fontsize={\"xticks\": 5, \"yticks\": 5},\n",
    "            tools=['tap'], \n",
    "            shared_axes=False,\n",
    "            scalebar=True, \n",
    "            scalebar_range='x',\n",
    "            scalebar_location=\"top_left\",\n",
    "            scalebar_unit=('bp'),))))\n",
    "p2 = pn.Column(showloc2, pn.pane.HoloViews(dynspread(rasterize(pafcis2).relabel(\"%s Cis HiC interactions\" % haps[1]).opts(\n",
    "            cmap=\"inferno\",\n",
    "            cnorm=\"log\",\n",
    "            colorbar=True,\n",
    "            width=pcwidth,\n",
    "            height=pcwidth,\n",
    "            xticks=qtic2,\n",
    "            yticks=qtic2,\n",
    "            xrotation=45,\n",
    "            fontsize={\"xticks\": 5, \"yticks\": 5},\n",
    "            tools=['tap'],\n",
    "            shared_axes=False,\n",
    "            scalebar=True, \n",
    "            scalebar_range='x',\n",
    "            scalebar_location=\"top_left\",\n",
    "            scalebar_unit=('bp'),\n",
    "            ))))\n",
    "\n",
    "p12 = pn.Row(p1,p2)\n",
    "\n",
    "p3 = pn.Row(pn.Column(showloctrans, pn.pane.HoloViews(dynspread(rasterize(paftrans).relabel(\"%s/%s Trans HiC interactions\" % (haps[0], haps[1])).opts(\n",
    "            cmap=\"inferno\",\n",
    "            cnorm=\"log\",\n",
    "            colorbar=True,\n",
    "            width=ptwidth,\n",
    "            height=ptwidth,\n",
    "            xticks=qtic1,\n",
    "            yticks=qtic2,\n",
    "            xrotation=45,\n",
    "            shared_axes=False,\n",
    "            fontsize={\"xticks\": 5, \"yticks\": 5},\n",
    "            tools=['tap'],\n",
    "            scalebar=True, \n",
    "            scalebar_range='x',\n",
    "            scalebar_location=\"top_left\",\n",
    "            scalebar_unit=('bp'))))))\n",
    "\n",
    "pn.panel(pn.Column(p12, p3)).servable(title='Artic Ground Squirrel Arima HiC VGP interactive heatmap')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "914671b9-5edc-41bb-a48a-077e77fbbf5e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "520af4d4-56f2-41ba-9161-46c017c539dc",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eaf32112-fd6c-4629-91cd-5309ea6a87ba",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "61ed8cf7-2919-42fa-bee2-6da0baaa7c4a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8c5e583b-de78-49fd-b6b6-bec2109afcc3",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7d7806d7-178b-4494-b053-3186f828a822",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fc07d612-9405-4a39-97ed-1e800bd65e7c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bc89f5e9-2981-4f32-8444-71251c6ed6d1",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7f3798e8-71d2-4de5-9d4b-a1172a22a47c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a64b5c4e-6b24-46a3-95a7-1bb98130469b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cde52a4a-1b40-462c-bcd4-8ceee6f17190",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a55afa84-0cf7-4b17-af09-99ddb366ab82",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fcbf0062-7d87-4d75-907a-232a2308315d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "15fd702d-507d-42eb-a5fd-7d17f55d0f26",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
